### PR TITLE
[FLINK-15411][table-planner-blink] Fix prune partition on DATE/TIME/TIMESTAMP columns

### PR DESCRIPTION
**Pull request by JingsongLi**
## What is the purpose of the change

When use DATE/TIME/TIMESTAMP partition columns, planner will not do partition pruning, that lead to read all partitions.

## Brief change log

- Fix comparison of timestamp with local time zone
- Fix prune partition on DATE/TIME/TIMESTAMP columns
- Fix listPartitions in HiveCatalog & Add time prune partition tests

## Verifying this change

- HiveTableSourceTest.testPartitionFilterDateTimestamp
- PartitionPruneTest.testTimePrunePartitions

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no

### CODE REVIEW

**Comment by JingsongLi** [comment]
CC:
@wuchong for blink part.
@lirui-apache for hive part.

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/TemporalTypesTest.scala [hunk 1]
```diff
@@ -783,6 +783,10 @@ class TemporalTypesTest extends ExpressionTestBase {
     testSqlApi(timestampTz("2018-03-14 19:01:02.123"), "2018-03-14 19:01:02.123")
     testSqlApi(timestampTz("2018-03-14 19:00:00.010"), "2018-03-14 19:00:00.01")
 
+    testSqlApi(
+      timestampTz("2018-03-14 19:00:00.010") + " > " + timestampTz("2018-03-14 19:00:00.012"),
```
Comment (wuchong) [inline flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/TemporalTypesTest.scala]: It seems that this test doesn't reproduce the bug. Could use replace one of the constants to a field reference?

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalog.java [hunk 2]
```diff
@@ -953,7 +954,7 @@ private static CatalogPartitionSpec createPartitionSpec(String hivePartitionName
 		Map<String, String> spec = new HashMap<>(partKeyVals.length);
 		for (String keyVal : partKeyVals) {
 			String[] kv = keyVal.split("=");
-			spec.put(kv[0], kv[1]);
+			spec.put(unescapePathName(kv[0]), unescapePathName(kv[1]));
```
Comment (lirui-apache) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalog.java]: Not related to this PR, but maybe we need more tests for this, e.g. string partition columns with special characters.

Comment (JingsongLi) [reply flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalog.java]: I added test in `HiveTableSourceTest.testPartitionFilter`

**Comment by wuchong** [comment]
Merging...

**Pull request merged**